File: packages/core/src/node/middleware.ts
/**
 * Hono-shaped observability middleware.
 *
 * For each request:
 *   1. Read user/org/session from `c.get('auth')` if the host app populated it
 *      (or via a caller-supplied `contextProvider`) and attach to the scope.
 *   2. Add a "request" context (method, path, headers subset, request id).
 *   3. Wrap the handler chain in `withScope(...)` so any `captureException`
 *      fired from a downstream handler picks up that request's identity.
 *   4. On thrown errors, call `captureException` BEFORE re-throwing so Hono's
 *      onError still gets to render the response.
 *
 * Why no AsyncLocalStorage in v1: under Lambda, each invocation is single-
 * request, so the module-scope singleton stack from `core/scope.ts` is safe.
 * If we ever target a long-lived multi-tenant Node server we'll need to swap
 * scope.ts over to ALS — tracked as a follow-up under the SDK epic.
 *
 * Naming follows Hono's middleware conventions but the type is structurally
 * compatible (we don't import `hono` from a core package).
 */

import { Client } from '../client';
import { withScope } from '../scope';

interface HonoCtxLike {
    req: {
        method: string;
        path?: string;
        url?: string;
        header: (name: string) => string | undefined;
    };
    get: (key: string) => unknown;
    set?: (key: string, value: unknown) => void;
    res?: { status?: number };
}

type HonoNext = () => Promise<void>;

export interface ObservabilityMiddlewareOptions {
    /**
     * Pull user identity out of the Hono context. Default reads `c.get('auth')`
     * shape (matches `@smooai/auth`'s authMiddleware output). Return `null` to
     * skip — the middleware will still capture errors but with no user.
     */
    resolveUser?: (c: HonoCtxLike) => { id?: string; orgId?: string; sessionId?: string } | null;
    /**
     * Header names whose values are recorded on the request context. Defaults
     * to a conservative allowlist that's safe to send to the ingest backend.
     */
    requestHeaderAllowlist?: string[];
}

const DEFAULT_HEADER_ALLOWLIST = ['user-agent', 'referer', 'x-request-id', 'x-trace-id', 'x-correlation-id'];

function defaultResolveUser(c: HonoCtxLike): { id?: string; orgId?: string; sessionId?: string } | null {
    const auth = c.get('auth') as { userId?: string; orgId?: string; sessionId?: string } | undefined;
    if (!auth) return null;
    return { id: auth.userId, orgId: auth.orgId, sessionId: auth.sessionId };
}

export function observabilityMiddleware(opts: ObservabilityMiddlewareOptions = {}) {
    const resolveUser = opts.resolveUser ?? defaultResolveUser;
    const allowlist = (opts.requestHeaderAllowlist ?? DEFAULT_HEADER_ALLOWLIST).map((h) => h.toLowerCase());

    return async function smooObservabilityMiddleware(c: HonoCtxLike, next: HonoNext): Promise<void> {
        if (!Client._isInitialized()) {
            // SDK not initialized — pass through so the host app isn't blocked.
            await next();
            return;
        }

        await withScope(async (scope) => {
            try {
                const user = resolveUser(c);
                if (user) scope.setUser(user);

                const headers: Record<string, string> = {};
                for (const name of allowlist) {
                    const v = c.req.header(name);
                    if (v) headers[name] = v;
                }
                scope.setContext('request', {
                    method: c.req.method,
                    path: c.req.path ?? c.req.url ?? '',
                    headers,
                });
            } catch {
                /* swallow — scope hydration must not break the request */
            }

            try {
                await next();
            } catch (err) {
                try {
                    Client.captureException(err, { tags: { source: 'hono.middleware' } });
                } catch {
                    /* swallow */
                }
                throw err;
            }
        });
    };
}

File: go/middleware.go
package observability

import (
	"fmt"
	"net/http"
)

// net/http middleware — the Go framework integration, analogous to the Hono
// middleware in node/middleware.ts. For each request it:
//  1. Establishes a fresh request-scoped Scope on the request context.
//  2. Resolves user identity (via ResolveUser) and a "request" context block.
//  3. Recovers from panics in the downstream handler, capturing them as
//     exceptions (status 500) before re-panicking so the host's own recovery
//     can still run.
//
// Fiber and Gin integrations live in their own modules (go/fiber, go/gin) as
// thin adapters over the same Scope + CaptureException primitives. Echo is still
// deferred — see the gap note in the README.

// MiddlewareOptions configures Middleware.
type MiddlewareOptions struct {
	// Client to capture on. Defaults to the package Default client.
	Client *Client
	// ResolveUser extracts user identity from the request. Return nil to skip.
	ResolveUser func(r *http.Request) *User
	// RequestHeaderAllowlist names headers recorded on the request context.
	// Defaults to a conservative, safe-to-send set.
	RequestHeaderAllowlist []string
	// SwallowPanics, when true, makes the middleware capture a downstream
	// panic, write a 500, and NOT re-panic. The default (false) re-panics
	// after capturing so the host's own recovery middleware still runs.
	SwallowPanics bool
}

var defaultHeaderAllowlist = []string{"user-agent", "referer", "x-request-id", "x-trace-id", "x-correlation-id"}

// Middleware returns a net/http middleware (func(http.Handler) http.Handler).
func Middleware(opts MiddlewareOptions) func(http.Handler) http.Handler {
	client := opts.Client
	if client == nil {
		client = Default
	}
	allowlist := opts.RequestHeaderAllowlist
	if allowlist == nil {
		allowlist = defaultHeaderAllowlist
	}
	rethrow := !opts.SwallowPanics

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if !client.IsInitialized() {
				next.ServeHTTP(w, r)
				return
			}

			scope := NewScope()
			ctx := ContextWithScope(r.Context(), scope)
			r = r.WithContext(ctx)

			func() {
				defer recoverSilently()
				if opts.ResolveUser != nil {
					if u := opts.ResolveUser(r); u != nil {
						scope.SetUser(u)
					}
				}
				headers := map[string]string{}
				for _, name := range allowlist {
					if v := r.Header.Get(name); v != "" {
						headers[name] = v
					}
				}
				scope.SetContext("request", map[string]any{
					"method":  r.Method,
					"path":    r.URL.Path,
					"headers": headers,
				})
			}()

			defer func() {
				if rec := recover(); rec != nil {
					err := panicToError(rec)
					client.CaptureExceptionOnSpan(ctx, err, map[string]string{"source": "http.middleware"})
					if rethrow {
						panic(rec)
					}
					w.WriteHeader(http.StatusInternalServerError)
				}
			}()

			next.ServeHTTP(w, r)
		})
	}
}

// NewMiddleware is the convenience constructor — captures panics then re-panics
// so the host's recovery still runs.
func NewMiddleware(client *Client, resolveUser func(r *http.Request) *User) func(http.Handler) http.Handler {
	return Middleware(MiddlewareOptions{
		Client:      client,
		ResolveUser: resolveUser,
	})
}

func panicToError(rec any) error {
	if err, ok := rec.(error); ok {
		return err
	}
	return fmt.Errorf("panic: %v", rec)
}
